// twilio-sync/lib/listitem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @class
 * @classdesc Represents an individual element in a Sync List.
 * @alias ListItem
 * @property {Number} index The index, within the containing List, of this item. This index is stable;
 * even if lower-indexed Items are removed, this index will remain as is.
 * @property {Object} data The contents of the item.
 * @property {Date} dateUpdated Date when the List Item was last updated.
 */
class ListItem {
    /**
     * @private
     * @constructor
     * @param {Object} data Item descriptor
     * @param {Number} data.index Item identifier
     * @param {String} data.uri Item URI
     * @param {Object} data.data Item data
     */
    constructor(data) {
        this.descriptor = data;
    }
    get uri() {
        return this.descriptor.uri;
    }
    get revision() {
        return this.descriptor.revision;
    }
    get lastEventId() {
        return this.descriptor.lastEventId;
    }
    get dateUpdated() {
        return this.descriptor.dateUpdated;
    }
    get dateExpires() {
        return this.descriptor.dateExpires;
    }
    get index() {
        return this.descriptor.index;
    }
    get data() {
        return this.descriptor.data;
    }
    /**
     * @private
     */
    update(eventId, revision, data, dateUpdated) {
        this.descriptor.lastEventId = eventId;
        this.descriptor.revision = revision;
        this.descriptor.data = data;
        this.descriptor.dateUpdated = dateUpdated;
        return this;
    }
    /**
     * @private
     */
    updateDateExpires(dateExpires) {
        this.descriptor.dateExpires = dateExpires;
    }
}
exports.ListItem = ListItem;
exports.default = ListItem;
