// twilio-notifications/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.log = exports.Logger = void 0;
const logger = require("loglevel");
function prepareLine(prefix, args) {
    return [`${new Date().toISOString()} Notifications ${prefix}:`].concat(Array.from(args));
}
class Logger {
    constructor() {
        this.prefix = '';
    }
    static scope(prefix = '') {
        // TBD this.prefix += ' ' + prefix;
        return new Logger();
    }
    setLevel(level) {
        logger.setLevel(level);
    }
    trace(...args) { logger.debug.apply(null, prepareLine('T' + this.prefix, args)); }
    debug(...args) { logger.debug.apply(null, prepareLine('D' + this.prefix, args)); }
    info(...args) { logger.info.apply(null, prepareLine('I' + this.prefix, args)); }
    warn(...args) { logger.warn.apply(null, prepareLine('W' + this.prefix, args)); }
    error(...args) { logger.error.apply(null, prepareLine('E' + this.prefix, args)); }
}
exports.Logger = Logger;
let logInstance = Logger.scope();
exports.log = logInstance;
